spec/lib/modules/filter/query_filter_encoded_spec.rb

```diff
@@ -0,0 +1,229 @@
+# frozen_string_literal: true
+
+describe Filter::Query do
+  create_entire_hierarchy
+
+  include SqlHelpers::Example
+
+  def compare_filter_sql(filter, sql_result)
+    filter_query = create_filter(filter)
+    comparison_sql(filter_query.query_full.to_sql, sql_result)
+    filter_query
+  end
+
+  def create_filter(params)
+    Filter::Query.new(
+      params,
+      AudioRecording.all,
+      AudioRecording,
+      AudioRecording.filter_settings
+    )
+  end
+
+  let(:test_filter) {
+    '{"filter":{"regions.id":{"eq":11}},"sorting":{"order_by":"recorded_date","direction":"desc"},"paging":{"items":25},"projection":{"include":["id","recorded_date","sites.name","site_id","canonical_file_name"]}}'
+  }
+
+  let(:test_filter_encoded) {
+    #  Base64.urlsafe_encode64(test_filter)
+    # Note: padding characters were removed as most encoders do not include them for base64url
+    'eyJmaWx0ZXIiOnsicmVnaW9ucy5pZCI6eyJlcSI6MTF9fSwic29ydGluZyI6eyJvcmRlcl9ieSI6InJlY29yZGVkX2RhdGUiLCJkaXJlY3Rpb24iOiJkZXNjIn0sInBhZ2luZyI6eyJpdGVtcyI6MjV9LCJwcm9qZWN0aW9uIjp7ImluY2x1ZGUiOlsiaWQiLCJyZWNvcmRlZF9kYXRlIiwic2l0ZXMubmFtZSIsInNpdGVfaWQiLCJjYW5vbmljYWxfZmlsZV9uYW1lIl19fQ'
+  }
+
+  context 'filter_encoded query string parameter' do
+    context 'errors' do
+      it 'rejects invalid base64 strings (an arbitrary string)' do
+        expect {
+          create_filter({
+            filter_encoded: 'banana'
+          }).query_full
+        }.to raise_error(
+          CustomErrors::FilterArgumentError, 'filter_encoded was not a valid RFC 4648 base64url string'
+        )
+      end
+
+      it 'rejects invalid json (an arbitrary string)' do
+        expect {
+          create_filter({
+            filter_encoded: '0o0o0o0o'
+          }).query_full
+        }.to raise_error(
+          CustomErrors::FilterArgumentError,
+          /filter_encoded was not a valid JSON payload: unexpected token at '.*'/
+        )
+      end
+
+      it 'rejects invalid base64URL encoding' do
+        expect {
+          create_filter({
+            filter_encoded: 'banana!'
+          }).query_full
+        }.to raise_error(
+          CustomErrors::FilterArgumentError, 'filter_encoded was not a valid RFC 4648 base64url string'
+        )
+      end
+
+      it 'rejects truncated values' do
+        expect {
+          create_filter({
+            filter_encoded: test_filter_encoded[0..-3]
+          }).query_full
+        }.to raise_error(
+          CustomErrors::FilterArgumentError,
+          /filter_encoded was not a valid JSON payload: unexpected token at '.*'/
+        )
+      end
+    end
+
+    context 'when decoding' do
+      it 'works as intended' do
```

Comment:
Would like to see non latin1 characters used in the test text as well (ie. russian language or symbols)